사장님 가게 생성

TIP-Santa opened the pull request:
#### 시나리오

---

가게 생성 시 권한 조회(OWNER)
가게 생성 시 동일한 이름의 가게를 조회
가게 생성 시 해당 OWNER의 가게가 3개 이상일 경우 생성 불가

---
#### 가게 생성 테스트
---
1. 가게 생성 성공
 - 동일한 가게 이름 없음
 - 현재 생성된 가게 수가 3개 미만

2. 가게 생성 실패 케이스
 - 선택한 카테고리가 없는 경우
 - 동일한 가게 이름이 있는 경우
 - 현재 생성된 가게 수가 3개 이상인 경우

TIP-Santa committed:
See issue #9 feat(store-controller) : 가게 생성 기능 추가

TIP-Santa committed:
See issue #9 feat(store-create-req-dto) : store-create-req-dto 추가

TIP-Santa committed:
See issue #9 feat(store-service) : 가게 생성 시 가게이름 중복검사 추가

TIP-Santa committed:
See issue #9 feat(error-code) : 가게이름 중복검사 에러코드 추가

TIP-Santa committed:
See issue #9 feat(store-repository) : 가게이름 중복검사 추가

TIP-Santa committed:
See issue #9 feat(store-create-req-dto) : Getter, Setter 어노테이션 추가

TIP-Santa committed:
See issue #9 feat(store-controller) : Token 추출 기능 추가

TIP-Santa committed:
See issue #9 feat(error-code) : 가게 수 초과 에러코드 추가

TIP-Santa committed:
See issue #9 feat(store-repository) : 가게 수 카운트 기능 추가

TIP-Santa committed:
See issue #9 feat(store-service) : 가게 수 검사 기능 추가

TIP-Santa committed:
See issue #9 feat(store-service) : OWNER 권한 확인 추가

TIP-Santa committed:
See issue #9 feat(error-code) : OWNER 검색 실패 코드 추가

TIP-Santa committed:
See issue #9 feat(store) : builder 어노테이션 추가

TIP-Santa committed:
See issue #9 feat(store-create-req-dto) : closeTime > endTime 변경

TIP-Santa committed:
See issue #9 feat(store-create-resp-dto) : storeCreateRespDto 구성 추가

TIP-Santa committed:
See issue #9 feat(store-service) : 가게 생성 기능 추가

TIP-Santa committed:
See issue #9 test(store-service) : storeServiceTest 성공 테스트

TIP-Santa committed:
See issue #9 build(build-gradle) : storeServiceTest 관련 설정

TIP-Santa committed:
See issue #9 build(application-test-yaml) : storeServiceTest 관련 설정

TIP-Santa committed:
See issue #9 test(store-service) : 실패 테스트 - 동일한 가게이름이 존재하는 경우

TIP-Santa committed:
See issue #9 test(store-service) : 실패 테스트 - 가게의 생성 수 제한

TIP-Santa committed:
See issue #9 chore(application.yaml) : application-test.yaml MySql 설정

TIP-Santa committed:
See issue #9 chore(store-service) : 문서 포멧팅

TIP-Santa committed:
See issue #9 refactor(store-controller) : ownerId 획득 방법 변경

TIP-Santa committed:
merge

TIP-Santa committed:
merge

TIP-Santa committed:
feat(error-code) : 카테고리 not found error-code 추가

TIP-Santa committed:
Revert "feat(error-code) : 카테고리 not found error-code 추가"

This reverts commit 966e32e5841f84d9f445de6eb19f469225039d55.

TIP-Santa committed:
feat(error-code) : 카테고리 not found error-code 추가 #9

TIP-Santa committed:
feat(store-category) : builder 어노테이션 추가 #9

TIP-Santa committed:
feat(store-create-req-dto) : categoryIds 추가

TIP-Santa committed:
feat(store-service) : 카테고리 저장기능 추가

TIP-Santa committed:
merge

TIP-Santa committed:
merge

TIP-Santa committed:
merge

TIP-Santa committed:
merge

TIP-Santa committed:
test(store-service-test) : 카테고리 추가 관련 성공 테스트

TIP-Santa committed:
test(store-service-test) : 카테고리 추가 관련 실패 테스트 - 카테고리 ID가 없는 경우 #9

TIP-Santa committed:
merge

yaewon0411 commented on src/main/java/com/my/foody/domain/store/controller/StoreController.java:
```diff
@@ -1,4 +1,32 @@
+
+    @RequireAuth(userType = UserType.OWNER)
+    @PostMapping("/stores")
+    public ResponseEntity<ApiResult<StoreCreateRespDto>> createStore(@RequestBody @Valid StoreCreateReqDto storeCreatereqDto, TokenSubject tokenSubject) {
```
TokenSubject 앞에 @CurrentUser 어노테이션이 붙어 있어야 합니다!

TIP-Santa replied:
감사합니다. 
@CurrentUser TokenSubject tokenSubject
수정했습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -1,4 +1,68 @@
+    private final CategoryRepository categoryRepository;
+    private final StoreCategoryRepository storeCategoryRepository;
+
+    public StoreCreateRespDto createStore(@Valid StoreCreateReqDto storeCreatereqDto, Long ownerId) {
```
여기에 @Valid를 다신 이유가 있나요?

TIP-Santa replied:
자동생성으로 같이 들어간 것 같습니다. 제거했습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -1,4 +1,68 @@
+    private final CategoryRepository categoryRepository;
+    private final StoreCategoryRepository storeCategoryRepository;
+
+    public StoreCreateRespDto createStore(@Valid StoreCreateReqDto storeCreatereqDto, Long ownerId) {
```
두 개 이상의 쓰기 작업을 하고 계시기 때문에 트랜잭션 어노테이션이 필요합니다

메서드에는 @Transactional을 걸어주시고
클래스 상단에서는 전역적으로 읽기 전용 트랜잭션 @Transactional(readOnly = true) 걸어주세요

TIP-Santa replied:
감사합니다.
@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
...
@Transactional
    public StoreCreateRespDto createStore(StoreCreateReqDto storeCreatereqDto, Long ownerId) {
수정했습니다.

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -1,4 +1,68 @@
+        Owner owner = ownerRepository.findById(ownerId).orElseThrow(() ->
+                new BusinessException(ErrorCode.OWNER_NOT_FOUND)
+        );
+        // 가게이름 중복 검사
+        if (storeRepository.existsByName(storeCreatereqDto.getName())) {
+            throw new BusinessException(ErrorCode.STORENAME_ALREADY_EXISTS);
+        }
+        // 사장님 가게 3개 이상 생성 불가
+        if (storeRepository.countByOwnerId(ownerId) >= 3) {
+            throw new BusinessException(ErrorCode.HAVE_FULL_STORE);
```
👍👍👍
유효성 검증하는 부분은 따로 메서드로 빼시는게 더 가독성에 좋을 거 같아요

ex) validateStoreCreation(storeCreaterqDto, ownerId) {...//가게 이름 중복 검사  //3개 이상 생성 불가 검사}

TIP-Santa replied:
감사합니다.
```
// 유효성 검사
        validateCreateStore(storeCreatereqDto, ownerId);
```
유효성 검사를 분리하였습니다.
```
public void validateCreateStore(StoreCreateReqDto storeCreatereqDto, Long ownerId) {
        // 가게이름 중복 검사
        if (storeRepository.existsByName(storeCreatereqDto.getName())) {
            throw new BusinessException(ErrorCode.STORENAME_ALREADY_EXISTS);
        }
        // 사장님 가게 3개 이상 생성 불가
        if (storeRepository.countByOwnerId(ownerId) >= 3) {
            throw new BusinessException(ErrorCode.HAVE_FULL_STORE);
        }
    }
```

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -1,4 +1,68 @@
+            throw new BusinessException(ErrorCode.HAVE_FULL_STORE);
+        }
+        // 가게 생성
+        Store store = Store.builder()
+                .name(storeCreatereqDto.getName())
+                .owner(owner)
+                .description(storeCreatereqDto.getDescription())
+                .contact(storeCreatereqDto.getContact())
+                .minOrderAmount(storeCreatereqDto.getMinOrderAmount())
+                .openTime(storeCreatereqDto.getOpenTime())
+                .endTime(storeCreatereqDto.getEndTime())
+                .build();
```
store 생성을 위한 빌더가 너무 크기 때문에 DTO에서 수행하는 방식을 제안합니다
ex)

```
//DTO 클래스 내
    public Store toEntity(Owner owner){
        return Store.builder()
                .name(name)
                .owner(owner)
                .description(description)
                .contact(contact)
                .minOrderAmount(minOrderAmount)
                .openTime(openTime)
                .endTime(endTime)
                .build();
    }

//서비스 단

        // 가게 생성
        Store store = storeCreatereqDto.toEntity(owner);
        storeRepository.save(store);
```

TIP-Santa replied:
감사합니다.
빌더를 DTO에서 수행하도록 수정하였습니다.
```
@Builder
    public Store toBuild(Owner owner) {
        return Store.builder()
                .name(name)
                .owner(owner)
                .description(description)
                .contact(contact)
                .minOrderAmount(minOrderAmount)
                .openTime(openTime)
                .endTime(endTime)
                .build();
    }
```
```
Store store = storeCreateReqDto.toBuild(owner);
```

TIP-Santa replied:
감사합니다.
빌더를 DTO에서 수행하도록 수정하였습니다.
```
@Builder
    public Store toBuild(Owner owner) {
        return Store.builder()
                .name(name)
                .owner(owner)
                .description(description)
                .contact(contact)
                .minOrderAmount(minOrderAmount)
                .openTime(openTime)
                .endTime(endTime)
                .build();
    }
```
```
Store store = storeCreateReqDto.toEntity(owner);
```

yaewon0411 commented on src/main/java/com/my/foody/domain/store/dto/req/StoreCreateReqDto.java:
```diff
@@ -1,4 +1,22 @@
+@Setter
+@NoArgsConstructor
 public class StoreCreateReqDto {
+    private String name;
+    private String description;
+    private String contact;
+    private Long minOrderAmount;
+    private LocalTime openTime;
+    private LocalTime endTime;
+    private boolean isDeleted;
+    private List<Long> categoryIds;
```
각 필드에 대한 유효성 검증이 필요해 보입니다

TIP-Santa replied:
앗... 초기에 설정했다가 브랜치 삭제하고 다시 만들면서 잊고있었네요... 감사합니다!
수정했습니다.
```
@NotBlank(message = "가게 이름를 입력해주세요")
    @Size(min = 1, max = 20, message = "가게 이름은 최대 20자까지 입력할 수 있습니다.")
    private String name;
    @NotBlank(message = "가게 설명을 입력해주세요")
    private String description;
    @NotBlank(message = "가게의 전화번호를 입력해주세요")
    @Size(min=1, max=15, message = "전화번호는 최대 15자까지 입력할 수 있습니다.")
    private String contact;
    @NotBlank(message = "최소 주문금액을 입력해주세요")
    @Pattern(regexp = "^[0-9]+$", message = "최소 주문금액은 숫자만 입력할 수 있습니다.")
    private Long minOrderAmount;
    @NotBlank(message = "가게 오픈시간을 입력해주세요")
    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "오픈시간은 시간으로 입력해주세요. ex) 11:00")
    private LocalTime openTime;
    @NotBlank(message = "가게 마감시간을 입력해주세요")
    @Pattern(regexp = "^([01]?[0-9]|2[0-3]):([0-5][0-9])$", message = "마감시간은 시간으로 입력해주세요. ex) 23:00")
    private LocalTime endTime;
    @NotBlank
    private boolean isDeleted;
    @NotBlank(message = "가게의 카테고리를 입력해주세요")
    @Pattern(regexp = "^[0-9]+$", message = "카테고리에 맞는 숫자로 입력해주세요.")
    private List<Long> categoryIds;
```

yaewon0411 commented on src/main/java/com/my/foody/domain/store/service/StoreService.java:
```diff
@@ -1,4 +1,68 @@
+                .endTime(storeCreatereqDto.getEndTime())
+                .build();
+        storeRepository.save(store);
+        // 카테고리 저장
+        List<Long> categoryIds = storeCreatereqDto.getCategoryIds();
+        for (Long categoryId : categoryIds) {
+            Category category = categoryRepository.findById(categoryId).orElseThrow(() ->
+                    new BusinessException(ErrorCode.CATEGORY_NOT_FOUND)
+            );
+            StoreCategory storeCategory = StoreCategory.builder()
+                    .store(store)
+                    .category(category)
+                    .build();
+            storeCategoryRepository.save(storeCategory);
+        }
```
카테고리 아이디 검증 좋습니다👍👍👍
카테고리 저장도 가독성을 위해서 별도 메서드로 분리하면 좋을 거 같네요

TIP-Santa replied:
감사합니다.
카테고리 저장 분리하였습니다.
```
// 카테고리 저장
        saveCategory(storeCreateReqDto, store);
```
```
// 카테고리 저장
    public void saveCategory(StoreCreateReqDto storeCreateReqDto, Store store) {
        List<Long> categoryIds = storeCreateReqDto.getCategoryIds();
        for (Long categoryId : categoryIds) {
            Category category = categoryRepository.findById(categoryId).orElseThrow(() ->
                    new BusinessException(ErrorCode.CATEGORY_NOT_FOUND)
            );
            StoreCategory storeCategory = StoreCategory.builder()
                    .store(store)
                    .category(category)
                    .build();
            storeCategoryRepository.save(storeCategory);
        }
    }
```

yaewon0411 reviewed (commented):
테스트 케이스 전부 통과했습니다!
고민 많이 하신 흔적이 보였습니다👍👍

gbognon25 reviewed (dismissed):
확인했습니다.

TIP-Santa committed:
refactor(store-controller) : TokenSubject 앞에 @CurrentUser 추가 #9

TIP-Santa committed:
refactor(store-service) : 리뷰 반영 #9
- @Valid 제거
- @Transactional(readOnly = true) 추가
- @Transactional 추가
- 유효성 검증 분리
- store 생성 builder를 dto에서 하도록 변경
- 카테고리 저장 분리

TIP-Santa committed:
refactor(store-create-resp-dto) : 어노테이션 추가

TIP-Santa committed:
refactor(store-create-resp-dto) : 리뷰 반영 #9
- 필드의 유효성 검사 추가
- Builder를 dto에서 수행하도록 변경

TIP-Santa committed:
feat(storeController) : 가게 조회 기능 구현 #30

TIP-Santa committed:
merge

TIP-Santa committed:
Update StoreService.java

미완성 코드 제거

TIP-Santa committed:
미완성 코드 제거

미완성 코드 제거

shushuaka commented:
확인 완료!

yaewon0411 commented:
테스트 케이스도 전부 통과되는 거 확인했습니다!
리뷰 반영 많이 해주셔서 감사해요
메서드 분리도 너무 좋고 유효성 검사도 세밀하게 되어있어서 좋았습니다

yaewon0411 reviewed (dismissed):
👍

TIP-Santa committed:
merge

TIP-Santa committed:
merge

TIP-Santa committed:
merge

TIP-Santa committed:
refactor(SotreColtroller) : 미완성 코드 정리

gbognon25 reviewed (approved):
확인했습니다~
